### ISSUE TITLE
[Bug]: Order of parameter properties - Discrepency between typescript and bable 

### ISSUE THREAD

**Issue by elpddev** [babel/babel#13779]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

@rollup/plugin-babel

### Input code

Original Code: 

```js
class Dashboard {
    name = 'stats' + this.amount;

    constructor(
        readonly amount
    ) { 
    }
}
```

Babel Emit:

```js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Dashboard {
  constructor(amount) {
    _defineProperty(this, "name", 'stats' + this.amount);

    this.amount = amount;
  }
}
```

Typescript emits:

```js
class Dashboard {
    constructor(amount) {
        this.amount = amount;
        this.name = 'stats' + this.amount;
    }
}
```

### Configuration file name

babel.config.js

### Configuration

For reproduction I've used the Babel online repl with Env stage 2 preset and Typescript preset. 
[link](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=MYGwhgzhAEAikAsBGB7MAnAJtA3gKGkOgDswBbAU2gF5oByCAFzEYjugGppGEBLCAHTkUAV2KMA3HgJFgKYk3QjgjFOgAUMokXQUwmeSACe0YWMZboASlzRLAXzz2gA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2%2Ctypescript&prettier=false&targets=&version=7.15.7&externalPlugins=&assumptions=%7B%7D)

For typescript reproduction I've used Typescript playground.
[link](https://www.typescriptlang.org/play?ssl=9&ssc=2&pln=1&pc=1#code/MYGwhgzhAEAikAsBGB7MAnAJtA3gKGkOgDswBbAU2gF5oByCAFzEYjugGppGEBLCAHTkUAV2KMA3HgJFgKYk3QjgjFOgAUMokXQUwmeSACe0YWMYAuEiLJIK6ADRboASlzTtAXzyegA)

### Current and expected behavior

Current Behavior - Babel puts parameter properties in the constructor after class properties assignment. Typescript puts them before.

Expected Behavior - Its a discrepancy. Not sure if Babel is the responsible or Typescript. Typescript behavior is more preferable to me because it makes writing shorter for me.

### Environment

- Babel - 7.15.7
- Typescript - 4.4.2

### Possible solution

Not sure. I've found some issues talking about plugin orders but I'm not sure.

**Comment by nicolo-ribaudo** [comment]
When comparing TS with Babel, you should always enable the `useDefineForClassFields` TS option: it's the option that makes TS class fields like JS class fields, which is what Babel does and what TC will do in a future version.

After enabling that option, if you target ES2015-ES2021 it will first initialize `this.amount` and then `this.name`. However, if you target ESNext (and soon ES2022, when it's supported), it first initialized `this.name` and then `this.amount`.

I suggest reporting this inconsistency to the TS team: when they'll decide which is the correct order we'll do the same.

**Comment by fedeci** [comment]
From developer.mozilla.org:
> Public instance fields are added with Object.defineProperty() either at construction time in the base class (before the constructor body runs), or just after super() returns in a subclass.

**Comment by nicolo-ribaudo** [comment]
Yeah, but MDN doesn't define parameter properties, which are a TS extension.

**Comment by Lyoko-Jeremie** [comment]
when i try to add babel to a typescript project, i face the same issue.

when use code like follow:

```ts

class B {

    constructor(
        public thisWin: Window,
    ) {
        console.log('B', [this, this.thisWin]);
        // now the `this.thisWin` is undefined
    }
    
    aFunc() {
        this.thisWin.xxx;   // <<=== Error: this.thisWin is undefined
    }

}

class A {

    constructor(
        public thisWin: Window,
    ) {
        console.log('A', [this, this.thisWin]);
    }

    b = new B(this.thisWin);

}

const a = new A(window);

a.b.aFunc();   // <<=== Error: this.thisWin is undefined

```

it will output if use babel

```
B [{...}, undefined ]
A [{...}, window ]
```

but only use typescript it will output :

```
B [{...}, window ]
A [{...}, window ]
```

**Comment by liuxingbaoyu** [comment]
Unfortunately we are waiting https://github.com/microsoft/TypeScript/issues/45995

**Comment by Lyoko-Jeremie** [comment]
@liuxingbaoyu 
I believe that initializing constructor parameters first should be the intuitive expected behavior. Users of TypeScript often come from backgrounds in other object-oriented programming languages such as C++, Java, C#, etc., where it is common for class internal members to be initialized using constructor parameters.

**Comment by nicolo-ribaudo** [comment]
The problem is that TypeScript behaves like Babel when you run it according to JS semantics for class properties, and differently when you run it with old pre-JS-classes semantics.

Before making any change we need to know how they intend to align the two implementations.

**Comment by Thulof** [comment]
When targeting an older version of ECMAScript during the build process, the behavior of Babel differs from other TypeScript compilers such as TSC or SWC. 

This discrepancy leads to a situation where no errors are raised during the development phase of the library, nor during the TypeScript checks in VSCode or CI. However, runtime errors emerge after the application is built. 

In my opinion, under similar conditions (for instance, when the build target is an earlier version of ES), Babel should maintain consistent results with those of other TypeScript compilers.

# Example

## Source Code:
```ts
class A {
  constructor(private a: number) {}
  private c = this.a + 2;
}
```

## Typescript Emit (Target: "ES5"): [TS Playground](https://www.typescriptlang.org/play?target=1&useDefineForClassFields=true#code/MYGwhgzhAECC0G8BQ1rAPYDsIBcBOArsDungBQAOeAlgG5g4Cm0YAXNJgQLYBGjeASkQBfFNCp0GzYNAC80HAAtqEAHRhoAamgAmANxJhQA)
```js
var A = /** @class */ (function () {
    function A(a) {
        this.a = a;
        this.c = this.a + 2;
    }
    return A;
}());
```

## SWC Emit (Target: "ES5"): [SWC Playground](https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAA0vOSSwuVnBUqOZSUEjOzysuKSpNLskv0igoyixLLElVSLRSyCvNTUot0lSorgWqgYknK9gqlGRkFuslKmgrGFlz1QIA4nq%2FcEkAAAA%3D&config=H4sIAAAAAAAAA1WPSw7DIAwF9zkF8rrbdtE79BCIOhERP9mOVBTl7oUE0maH3xszsA5KwcwGnmotxzIkTYx0ziXhHER%2FSgKSE7IhmwRuvRWu1agd4x5tRwOiaUKpW8j3hoOLkbHjLfM22DH%2FC030iZD5ClZUh8nhVTc0Jfj4XvayfaQ%2B9tA%2F4Ad12XkxWH71TaEFh%2B0LYuVI0xQBAAA%3D)
```js
var A = function A(a) {
    "use strict";
    _class_call_check(this, A);
    _define_property(this, "a", void 0);
    _define_property(this, "c", void 0);
    this.a = a;
    this.c = this.a + 2;
};
```

## Babel Emit (Target: "Chrome 64"): [Babel Playground](https://babeljs.io/repl#?browsers=chrome%2064&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1rAPYDsIBcBOArsDungBQAOeAlgG5g4Cm0YAXNJgQLYBGjeASkQBfFNCp0GzYNAC80HAAtqEAHRhoAamgAmANxJhQA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=true&targets=&version=7.24.4&externalPlugins=&assumptions=%7B%7D)
```js
class A {
  constructor(a) {
    _defineProperty(this, "c", this.a + 2);
    this.a = a;
  }
}
```
